File: src/main/java/br/com/controle/financeiro/controller/api/CardController.java
package br.com.controle.financeiro.controller.api;

import static org.springframework.hateoas.server.mvc.WebMvcLinkBuilder.linkTo;
import static org.springframework.hateoas.server.mvc.WebMvcLinkBuilder.methodOn;

import java.util.List;
import java.util.UUID;
import java.util.stream.Collectors;

import javax.validation.Valid;

import br.com.controle.financeiro.controller.api.linkbuilder.CardDTOResourceAssembler;
import br.com.controle.financeiro.model.dto.CardDTO;
import br.com.controle.financeiro.model.entity.Card;
import br.com.controle.financeiro.model.entity.Client;
import br.com.controle.financeiro.model.entity.Institution;
import br.com.controle.financeiro.model.entity.UserEntity;
import br.com.controle.financeiro.model.exception.CardNotFoundException;
import br.com.controle.financeiro.model.repository.CardRepository;
import br.com.controle.financeiro.model.repository.ClientRepository;
import br.com.controle.financeiro.model.repository.InstitutionRepository;
import br.com.controle.financeiro.service.UserService;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.hateoas.EntityModel;
import org.springframework.hateoas.CollectionModel;
import org.springframework.http.HttpStatus;
import org.springframework.http.MediaType;
import org.springframework.web.bind.annotation.DeleteMapping;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.PutMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.ResponseStatus;
import org.springframework.web.bind.annotation.RestController;

@RestController
@RequestMapping(value = "/api/card", produces = MediaType.APPLICATION_JSON_VALUE)
public class CardController {

    private static final Logger LOG = LoggerFactory.getLogger(CardController.class);

    private final CardRepository cardRepository;
    private final CardDTOResourceAssembler cardDTOResourceAssembler;
    private final ClientRepository clientRepository;
    private final InstitutionRepository institutionRepository;
    private final UserService userService;

    public CardController(final CardRepository cardRepository, final CardDTOResourceAssembler cardDTOResourceAssembler,
                          final ClientRepository clientRepository, final InstitutionRepository institutionRepository,
                          final UserService userService) {
        this.cardRepository = cardRepository;
        this.cardDTOResourceAssembler = cardDTOResourceAssembler;
        this.clientRepository = clientRepository;
        this.institutionRepository = institutionRepository;
        this.userService = userService;
    }

    @GetMapping
    public CollectionModel<EntityModel<CardDTO>> allCards() {
        LOG.debug("finding allCards");
        UserEntity owner = userService.getAuthenticatedUser();
        final List<EntityModel<CardDTO>> cards =
                cardRepository.findAllByOwner(owner).stream().map(CardDTO::fromCard).map(cardDTOResourceAssembler::toModel)
                              .collect(Collectors.toList());

        return new CollectionModel<>(cards, linkTo(methodOn(CardController.class).allCards()).withSelfRel());
    }

    @ResponseStatus(HttpStatus.CREATED)
    @PostMapping(consumes = { MediaType.APPLICATION_JSON_VALUE })
    public EntityModel<CardDTO> newCard(@RequestBody @Valid final CardDTO card) {
        LOG.debug("creating newCard");
        //TODO extract to service
        UserEntity owner = userService.getAuthenticatedUser();
        Client client = clientRepository.findByIdAndOwner(card.getClient(), owner).orElseThrow();
        Institution institution = institutionRepository.findById(card.getInstitution()).orElseThrow();
        CardDTO savedCard = CardDTO.fromCard(cardRepository.save(card.toCard(client, institution, owner)));
        return cardDTOResourceAssembler.toModel(savedCard);
    }

    @GetMapping(path = "/{id}")
    public EntityModel<CardDTO> oneCard(@PathVariable(value = "id") final UUID id) {
        LOG.debug("searching oneCard ${}", id);
        UserEntity owner = userService.getAuthenticatedUser();
        final Card savedCard = cardRepository.findByIdAndOwner(id, owner).orElseThrow(() -> new CardNotFoundException(id));
        return cardDTOResourceAssembler.toModel(CardDTO.fromCard(savedCard));
    }

    @PutMapping(path = "/{id}")
    public EntityModel<CardDTO> replaceCard(@RequestBody final CardDTO newCard, @PathVariable final UUID id) {
        LOG.info("replaceCard");
        //TODO verify DTO integrity
        final UserEntity owner = userService.getAuthenticatedUser();
        Card savedCard = cardRepository.findByIdAndOwner(id, owner).map(card -> {
            card.setName(newCard.getName());
            return cardRepository.save(card);
        }).orElseGet(() -> {
            //TODO extract to service
            Client c = clientRepository.findByIdAndOwner(newCard.getClient(), owner).orElseThrow();
            Institution i = institutionRepository.findById(newCard.getInstitution()).orElseThrow();
            newCard.setId(id);
            return cardRepository.save(newCard.toCard(c, i, owner));
        });

        return cardDTOResourceAssembler.toModel(CardDTO.fromCard(savedCard));
    }

    @ResponseStatus(HttpStatus.NO_CONTENT)
    @DeleteMapping(path = "/{id}")
    public void deleteCard(@PathVariable final UUID id) {
        LOG.debug("trying to deleteCard ${}", id);
        //TODO verify authenticated permission
        cardRepository.deleteById(id);
    }

}

File: src/main/java/br/com/controle/financeiro/controller/api/CardController.kt
package br.com.controle.financeiro.controller.api

import java.util.UUID
import br.com.controle.financeiro.model.entity.Institution
import br.com.controle.financeiro.model.entity.UserEntity
import br.com.controle.financeiro.model.dto.CardDTO
import com.fasterxml.jackson.annotation.JsonProperty
import br.com.controle.financeiro.model.dto.ClientDTO
import br.com.controle.financeiro.model.entity.BankAccount
import br.com.controle.financeiro.model.dto.BankAccountDTO
import br.com.controle.financeiro.model.dto.InstitutionDTO
import com.fasterxml.jackson.databind.annotation.JsonDeserialize
import com.fasterxml.jackson.datatype.jsr310.deser.LocalDateTimeDeserializer
import java.math.BigDecimal
import java.util.Locale
import br.com.controle.financeiro.model.dto.TransactionDTO
import org.springframework.data.jpa.domain.AbstractPersistable
import javax.persistence.FetchType
import javax.persistence.JoinColumn
import org.springframework.security.core.GrantedAuthority
import javax.persistence.GeneratedValue
import javax.persistence.GenerationType
import org.springframework.security.core.userdetails.UserDetails
import javax.persistence.ManyToMany
import java.lang.RuntimeException
import org.springframework.data.jpa.repository.JpaRepository
import org.springframework.data.repository.CrudRepository
import br.com.controle.financeiro.spring.conditionals.FirebaseConditionImpl
import org.springframework.context.annotation.ConditionContext
import org.springframework.core.type.AnnotatedTypeMetadata
import br.com.controle.financeiro.service.impl.UserServiceImpl
import br.com.controle.financeiro.service.UserService
import org.springframework.beans.factory.annotation.Autowired
import br.com.controle.financeiro.model.repository.UserRepository
import br.com.controle.financeiro.model.repository.RoleRepository
import org.springframework.security.core.userdetails.UsernameNotFoundException
import java.util.HashSet
import org.springframework.security.core.authority.SimpleGrantedAuthority
import org.springframework.security.access.annotation.Secured
import br.com.controle.financeiro.configuration.security.SecurityConfig.Roles
import br.com.controle.financeiro.service.UserService.RegisterUserInit
import org.springframework.security.core.context.SecurityContextHolder
import br.com.controle.financeiro.configuration.security.auth.firebase.FirebaseAuthenticationToken
import br.com.controle.financeiro.spring.conditionals.FirebaseCondition
import br.com.controle.financeiro.service.FirebaseParserService
import br.com.controle.financeiro.configuration.security.auth.firebase.FirebaseTokenHolder
import com.google.firebase.auth.AbstractFirebaseAuth
import java.lang.IllegalArgumentException
import com.google.firebase.auth.FirebaseToken
import br.com.controle.financeiro.service.exception.FirebaseTokenInvalidException
import org.springframework.security.authentication.BadCredentialsException
import org.springframework.security.core.userdetails.UserDetailsService
import org.springframework.hateoas.server.RepresentationModelAssembler
import org.springframework.hateoas.EntityModel
import org.springframework.hateoas.server.mvc.WebMvcLinkBuilder
import br.com.controle.financeiro.controller.api.CardController
import br.com.controle.financeiro.controller.api.ClientController
import br.com.controle.financeiro.controller.api.BankAccountController
import br.com.controle.financeiro.controller.api.InstitutionController
import br.com.controle.financeiro.controller.api.TransactionController
import br.com.controle.financeiro.model.repository.CardRepository
import br.com.controle.financeiro.controller.api.linkbuilder.CardDTOResourceAssembler
import br.com.controle.financeiro.model.repository.ClientRepository
import br.com.controle.financeiro.model.repository.InstitutionRepository
import org.springframework.hateoas.CollectionModel
import java.util.stream.Collectors
import br.com.controle.financeiro.model.exception.CardNotFoundException
import br.com.controle.financeiro.controller.api.linkbuilder.ClientDTOResourceAssembler
import br.com.controle.financeiro.model.exception.ClientNotFoundException
import br.com.controle.financeiro.model.repository.BankAccountRepository
import br.com.controle.financeiro.controller.api.linkbuilder.BankAccountDTOResourceAssembler
import br.com.controle.financeiro.model.exception.BankAccountNotFoundException
import br.com.controle.financeiro.controller.api.linkbuilder.InstitutionDTOResourceAssembler
import br.com.controle.financeiro.model.exception.InstitutionNotFoundException
import br.com.controle.financeiro.model.repository.TransactionRepository
import br.com.controle.financeiro.controller.api.linkbuilder.TransactionDTOResourceAssembler
import br.com.controle.financeiro.model.exception.TransactionNotFoundException
import org.springframework.web.servlet.ModelAndView
import org.springframework.web.servlet.mvc.method.annotation.ResponseEntityExceptionHandler
import org.springframework.web.context.request.WebRequest
import org.springframework.http.ResponseEntity
import org.springframework.dao.EmptyResultDataAccessException
import org.springframework.hateoas.client.LinkDiscoverers
import org.springframework.hateoas.client.LinkDiscoverer
import org.springframework.hateoas.mediatype.collectionjson.CollectionJsonLinkDiscoverer
import org.springframework.plugin.core.SimplePluginRegistry
import springfox.documentation.spring.web.plugins.Docket
import springfox.documentation.spi.DocumentationType
import springfox.documentation.builders.RequestHandlerSelectors
import springfox.documentation.builders.PathSelectors
import springfox.documentation.service.SecurityReference
import springfox.documentation.service.ApiKey
import org.springframework.web.filter.OncePerRequestFilter
import kotlin.Throws
import javax.servlet.ServletException
import java.io.IOException
import javax.servlet.http.HttpServletRequest
import javax.servlet.http.HttpServletResponse
import javax.servlet.FilterChain
import br.com.controle.financeiro.configuration.security.auth.firebase.FirebaseFilter
import java.lang.SecurityException
import org.springframework.security.authentication.AbstractAuthenticationToken
import org.apache.commons.lang3.builder.HashCodeBuilder
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.database.DatabaseReference
import com.google.firebase.database.FirebaseDatabase
import com.google.firebase.FirebaseApp
import br.com.controle.financeiro.configuration.security.FirebaseConfig
import com.google.auth.oauth2.GoogleCredentials
import com.google.firebase.FirebaseOptions
import java.io.ByteArrayInputStream
import org.springframework.security.config.annotation.method.configuration.EnableGlobalMethodSecurity
import org.springframework.security.config.annotation.method.configuration.GlobalMethodSecurityConfiguration
import br.com.controle.financeiro.configuration.security.auth.firebase.FirebaseAuthenticationProvider
import org.springframework.security.config.annotation.authentication.configuration.GlobalAuthenticationConfigurerAdapter
import org.springframework.security.config.annotation.authentication.builders.AuthenticationManagerBuilder
import org.springframework.security.crypto.password.PasswordEncoder
import org.springframework.security.crypto.factory.PasswordEncoderFactories
import org.springframework.security.config.annotation.web.configuration.WebSecurityConfigurerAdapter
import org.springframework.security.config.annotation.web.builders.WebSecurity
import org.springframework.security.config.annotation.web.builders.HttpSecurity
import org.springframework.security.web.authentication.www.BasicAuthenticationFilter
import br.com.controle.financeiro.configuration.security.SecurityConfig.ApplicationSecurity
import br.com.controle.financeiro.model.entity.Card
import org.slf4j.LoggerFactory
import org.springframework.security.config.http.SessionCreationPolicy
import org.springframework.boot.autoconfigure.SpringBootApplication
import springfox.documentation.swagger2.annotations.EnableSwagger2
import org.springframework.boot.autoconfigure.domain.EntityScan
import org.springframework.data.jpa.repository.config.EnableJpaRepositories
import kotlin.jvm.JvmStatic
import org.springframework.boot.SpringApplication
import org.springframework.http.HttpStatus
import org.springframework.http.MediaType
import org.springframework.web.bind.annotation.*
import java.util.function.Function
import javax.validation.Valid

@RestController
@RequestMapping(value = ["/api/card"], produces = [MediaType.APPLICATION_JSON_VALUE])
class CardController(
    private val cardRepository: CardRepository, private val cardDTOResourceAssembler: CardDTOResourceAssembler,
    private val clientRepository: ClientRepository, private val institutionRepository: InstitutionRepository,
    private val userService: UserService
) {
    @GetMapping
    fun allCards(): CollectionModel<EntityModel<CardDTO>> {
        LOG.debug("finding allCards")
        val owner = userService.authenticatedUser
        val cards = cardRepository.findAllByOwner(owner).stream().map { card: Card? -> CardDTO.fromCard(card) }
            .map { entity: CardDTO -> cardDTOResourceAssembler.toModel(entity) }
            .collect(Collectors.toList())
        return CollectionModel(
            cards, WebMvcLinkBuilder.linkTo(
                WebMvcLinkBuilder.methodOn(
                    CardController::class.java
                ).allCards()
            ).withSelfRel()
        )
    }

    @ResponseStatus(HttpStatus.CREATED)
    @PostMapping(consumes = [MediaType.APPLICATION_JSON_VALUE])
    fun newCard(@RequestBody card: @Valid CardDTO?): EntityModel<CardDTO> {
        LOG.debug("creating newCard")
        //TODO extract to service
        val owner = userService.authenticatedUser
        val client = clientRepository.findByIdAndOwner(card?.client, owner).orElseThrow()
        val institution = card?.institution?.let { institutionRepository.findById(it).orElseThrow() }!!
        val savedCard: CardDTO = CardDTO.fromCard(
            cardRepository.save(
                card.toCard(client, institution, owner)
            )
        )
        return cardDTOResourceAssembler.toModel(savedCard)
    }

    @GetMapping(path = ["/{id}"])
    fun oneCard(@PathVariable(value = "id") id: UUID?): EntityModel<CardDTO> {
        LOG.debug("searching oneCard \${}", id)
        val owner = userService.authenticatedUser
        val savedCard = cardRepository.findByIdAndOwner(id, owner).orElseThrow { CardNotFoundException(id) }
        return cardDTOResourceAssembler.toModel(CardDTO.fromCard(savedCard))
    }

    @PutMapping(path = ["/{id}"])
    fun replaceCard(@RequestBody newCard: CardDTO, @PathVariable id: UUID?): EntityModel<CardDTO> {
        LOG.info("replaceCard")
        //TODO verify DTO integrity
        val owner = userService.authenticatedUser
        val savedCard = cardRepository.findByIdAndOwner(id, owner).map { card: Card ->
            card.name = newCard.name
            cardRepository.save(card)
        }.orElseGet {

            //TODO extract to service
            val c = clientRepository.findByIdAndOwner(newCard.client, owner).orElseThrow()
            val i = institutionRepository.findById(newCard.institution).orElseThrow()!!
            newCard.id = id
            cardRepository.save(newCard.toCard(c, i, owner))
        }
        return cardDTOResourceAssembler.toModel(CardDTO.fromCard(savedCard))
    }

    @ResponseStatus(HttpStatus.NO_CONTENT)
    @DeleteMapping(path = ["/{id}"])
    fun deleteCard(@PathVariable id: UUID) {
        LOG.debug("trying to deleteCard \${}", id)
        //TODO verify authenticated permission
        cardRepository.deleteById(id)
    }

    companion object {
        private val LOG = LoggerFactory.getLogger(CardController::class.java)
    }
}